### PR TITLE
(#4070) NCIDS Dynamic List

### LINKED ISSUE #4070: Create NCIDS Dynamic List
## As a user on the News & Events landing page, I want to see a list of news releases, so that I can be updated on the latest cancer research news.

### Requirements
- Create a new NCIDS Dynamic List to display press releases on the News & Events page
- Use the usa-collections with the patch created in the NCIDS
- Below the press releases, there will be a View More link
- This component will match the Content Model

### Drupal DoR

- [ ] Content Model:
  - [x] Drafted
  - [x] Reviewed
  - [ ] Approved
- [ ] Acceptance Criteria:
  - [ ] Drafted
  - [ ] Reviewed
  - [ ] Approved
- [ ] YAML Created (make YAML for [News & Events Landing Page](https://www.figma.com/file/XH87Z6GvJ0MWoX16l0FzKo/News-%26-Events-Landing-Page?type=design&node-id=0-1&mode=design&t=asi2XMCLfH1MPRqE-0))

### Drupal DoD

- [ ] Drupal Governance Drafted
- [ ] Internal Testing
- [ ] Client Review

### Acceptance Criteria

```gherkin
Scenario: A content editor adds an NCIDS Dynamic List on a Home and Landing page
  Given a content editor is logged into the CMS
    And the user creates a new Home and Landing page content item
  When the user adds an NCIDS 2 Column Container paragraph to the Contents field
      And the user adds an NCIDS Dynamic List to the Main Contents field
  Then the paragraph fields will exactly match the content model

```

```gherkin
Scenario: User can add an NCIDS Dynamic List to a Home and Landing Page
  Given user is logged into the CMS
  When user creates a new "Home and Landing" content item
  Then the editing page for a "Home and Landing" content item displays
    And user fills out the following fields
      | fieldLabel                  | value                                                   | field_name                     |
      | Home and Landing Page Style | NCIDS without Title                                     | field_page_style               |
      | Pretty URL                  | ncids-dynamic-list-page                                 | field_pretty_url               |
      | Page Title                  | Automated Test NCIDS Dynamic List                       | title                          |
      | Browser Title               | Automated Test NCIDS 2 Dynamic List - Browser Title     | field_browser_title            |
      | Card Title                  | Automated Test NCIDS 2 Dynamic List - Card Title        | field_card_title               |
      | Meta Description            | Automated Test NCIDS 2 Dynamic ListMeta Description     | field_page_description         |
      | Feature Card Description    | Automated Test NCIDS 2 Dynamic List - Feature Card Desc | field_feature_card_description |
  When user selects "Add NCIDS 2 Column Container" from "Contents" dropdown
  Then "NCIDS 2 Column Container" section appears
  When user selects "Add NCIDS Dynamic List" from "Main Contents" dropdown
  Then "NCIDS Dynamic List" section appears
    And user fills out the following fields
      | fieldLabel         | value                      | field_name        |
      | List Title         | NCIDS Dynamic List Heading | field_list_title  |
      | Source View        | Press Releases             | ???               |
      | Display            | Home List Block            | ???               |
      | Include View Title | unchecked                  | ???               |
      | Items per page     |                            | ???               |
      | Pagination         | Default settings           | ???               |
      | Arguments          |                            | ???               |
  When user selects "Add Raw HTML Content" from "Secondary Contents" dropdown
  Then "NCIDS Raw HTML Content" section appears
    And user fills out the following fields
      | fieldLabel       | value                                                                                                                                                                                                    | field_name           |
      | Raw HTML Content | <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p><p>Curabitur gravida arcu ac tortor dignissim convallis aenean et.</p> | field_card_raw_html  |
    When user saves the content page
    Then user can see NCIDS 2 Column Container on the page
      And the Main Contents heading displays "NCIDS Dynamic List Heading"
      And the Main Contents dynamic list shows 5 items
      And the Secondary Contents displays "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Curabitur gravida arcu ac tortor dignissim convallis aenean et."
```

``` gherkin
Scenario: Analytics event for clicks on collections component
Given a user is on any page of cancer.gov with a collections component
When the user clicks on a link within the collections component
Then there should be an NCIDataLayer Other event 
  And the event has the name “LP:Collection:LinkClick”
  And the event has the the linkname “LP:Collection:LinkClick”
  And this event should have the following data:
    | key              | value                                                                                                                                                                                                                                                                                                                                                                                             |
    | location         | Body                                                                                                                                                                                                                                                                                                                                                                                              |
    | pageRows         | This should return the number of rows on the page overall                                                                                                                                                                                                                                                                                                                                                           |
    | pageRowIndex     | This should return the index of the row number on the page that was clicked |
    | rowItems         | This should return the total number of components in the row that is clicked|
    | rowItemIndex     | This should return the index of the component in the rows that was clicked |
    | componentType    | Collection |
    | componentTheme   | This should be “Light” or “Dark”. In cases where a theme option is not yet available, this match whatever the available theme color is.|
    | componentVariant | This value represents the component design options available for a component type. For Collections, this value should be set to “Standard”. |
    | title            | This value should represent the collection component overall. If there is a title field or some other value that can help make this more specific, that would be great. If not, we can pass “Collection” for this value. (Ideally, something like the heading above it would be great, but know that is a different component and not connected to this.)|
    | linkType         | This value can be set to “internal”, “external”, or “media” based on the destination/type of links clicked.|
    | linkText         | This should capture the title of the collection item clicked. In the case that a user clicks the button at the bottom of the component, this value should pass the text used on the button. This value should be truncated after 50 characters. |
    | linkArea         | This value represents the link or area that was clicked. For Collections, this value will be “Collection Item” and if the button at the end of collection is clicked pass “View All Button”. (If text and image on the collection item are individual clickable elements, this should be updated so that clicks on the title pass “Collection Item Title” and for the image “Collection Item Image”|
    | totalLinks       | This should pass the total number of collection items shown in the collections component. (Ex: if the component shows 4 items in the list, then pass this value as 4.) |
    | linkPosition     | This should pass the index of the link clicked in the component evaluated by counting the items top to bottom. (Ex: If the third item in the list is clicked, pass 3.)|
```

### Resources:
* [Mockups](https://www.figma.com/file/XH87Z6GvJ0MWoX16l0FzKo/News-%26-Events-Landing-Page?type=design&mode=design&t=nL9qJIUa7smfJMSF-0)
* [DS Collections component](https://github.com/NCIOCPL/ncids/issues/1149)
* [DS Headings component](https://github.com/NCIOCPL/ncids/issues/1150)


### Notes

### Prerequisites
- [ ] [NCIDS Headings](https://github.com/NCIOCPL/ncids/issues/1150)
- [ ] [NCIDS Collections](https://github.com/NCIOCPL/ncids/issues/1149)


### Tasks
- [ ] Create NCIDS Home List Block view (maybe? if so, let's rename this)


### Technical Notes
* [Content Model](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Paragraph-Type-NCIDS-Dynamic-List)

### PR DESCRIPTION
**Pull request by dev-rana-publicis**
Closes #4070
https://ncigovcdode669.prod.acquia-sites.com/

### CODE REVIEW

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
The javascript/css for the dynamic lists should be `cgdp-dynamic-list` and not `usa-collection`. The cgdp-dynamic-ist draws as a collection, but it is not JS and CSS for all collections. For example we have a "list" component that would almost certainly also render as a usa-collection. That list would be a `cgdp-list` or something. Both use usa-collection.

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
This is a review of c7abaa33b29dfc017339a47d9a213b0416e22cb3

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/cgdp-landing-raw-html.ts [hunk 1]
```diff
@@ -18,17 +18,36 @@ const rawHtmlClickTracker = (target: HTMLElement) => {
 		? pageRows.indexOf(parentRowContainer) + 1
 		: '_ERROR_';
 
+	/** All of the columns within the page row */
+	const pageRowCols = parentRowContainer
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/cgdp-landing-raw-html.ts:22]: This and all the raw HTML bits need to go into the 2 column commit.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/analytics/landing-page-contents-helper.ts [hunk 2]
```diff
@@ -54,10 +54,25 @@ export const landingClickTracker = (
 		? pageRows.indexOf(parentRowContainer) + 1
 		: '_ERROR_';
 
+	/** All of the columns within the page row */
+	const pageRowCols = parentRowContainer
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/core/analytics/landing-page-contents-helper.ts:58]: The core analytics here have to go into the 2 column layout commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-promo-block/__tests__/nci-promo-block.test.ts [hunk 3]
```diff
@@ -80,6 +80,8 @@ describe('cgdp-promo-block', () => {
 					location: 'Body',
 					pageRows: numberOfPromoBlockRows,
 					pageRowIndex: 7,
+					pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-promo-block/__tests__/nci-promo-block.test.ts:83]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-hero/__tests__/nci-hero.test.ts [hunk 4]
```diff
@@ -102,6 +102,8 @@ describe('NCI Hero', () => {
 			location: 'Body',
 			pageRows: 1,
 			pageRowIndex: 1,
+			pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-hero/__tests__/nci-hero.test.ts:105]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-cta-strip/__tests__/nci-cta-strip.test.ts [hunk 5]
```diff
@@ -57,6 +57,8 @@ describe('NCI CTA Strip', () => {
 				location: 'Body',
 				pageRows: 0,
 				pageRowIndex: '_ERROR_',
+				pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/nci-cta-strip/__tests__/nci-cta-strip.test.ts:60]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-inline-video/__tests__/cgdp-inline-video.test.ts [hunk 6]
```diff
@@ -57,6 +57,8 @@ describe('CGDP Inline Video', () => {
 				location: 'Body',
 				pageRows: 1,
 				pageRowIndex: 1,
+				pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-inline-video/__tests__/cgdp-inline-video.test.ts:60]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-guide-card-row/__tests__/cgdp-guide-card-row.test.ts [hunk 7]
```diff
@@ -71,6 +71,8 @@ describe('CGDP Guide Cards', () => {
 				location: 'Body',
 				pageRows: 1,
 				pageRowIndex: 1,
+				pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-guide-card-row/__tests__/cgdp-guide-card-row.test.ts:74]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-feature-card-row/__tests__/cgdp-feature-card.test.ts [hunk 8]
```diff
@@ -59,6 +59,8 @@ describe('CGDP Feature Cards', () => {
 				location: 'Body',
 				pageRows: 1,
 				pageRowIndex: 1,
+				pageRowCols: 0,
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-feature-card-row/__tests__/cgdp-feature-card.test.ts:62]: 2 col commit

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/__tests__/cgdp-landing-raw-html.test.ts [hunk 9]
```diff
@@ -7,6 +7,7 @@ import * as eddlUtil from '../../../core/analytics/eddl-util';
 
 import cgdpRawHtml from '../cgdp-landing-raw-html';
 import { cgdpLandingRawHtmlDom } from './cgdp-landing-raw-html.dom';
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/__tests__/cgdp-landing-raw-html.test.ts:9]: Add multiple rows to the cgdpLandingRawHtmlDom example if it does not have it.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/__tests__/cgdp-landing-raw-html-cols.dom.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
+export const cgdpRawHTMLRowCols = `
+<div class="usa-section usa-section--light" data-eddl-landing-row>
+    <div class="grid-container">
+        <div class="grid-row grid-gap-4">
+            <div class="tablet-lg:grid-col-8 ncids-column" data-eddl-landing-row-col>
+                <div class="ncids_dynamic_list" data-eddl-landing-item="collection">
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/lib/components/cgdp-landing-raw-html/__tests__/cgdp-landing-raw-html-cols.dom.ts:6]: This test dom needs to get updated to have:
1 raw html in column 1, and 2 raw html in column 2, and test cases to match.

There should be matching click handlers in the unit tests.

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
The pageRowCols stuff needs to be moved to the 2 col layout. Also the logic is not going to work right in figuring out what is a row in a col and what is a row in the landingContents. The approach was discussed in Slack last week, but raw html should always be something like:

`<div class="usa-section" data-eddl-landing-row>` no matter if in the landing contents or a row.

This needs updates to how you find pageRows and pageRowIndex (across the board) to be:

```
const allRows = Array.from(
	document.querySelectorAll('[data-eddl-landing-row]')
);
const nestedRows = Array.from(
	document.querySelectorAll('[data-eddl-landing-row] [data-eddl-landing-row]')
);
const pageRows = allRows.filter((r) => nestedRows.indexOf(r) === -1);

const col = target.closest('[data-eddl-landing-row-col]');
const parentRowContainer = col ? col.closest('[data-eddl-landing-row]') : target.closest('[data-eddl-landing-row]')
const pageRowIndex = parentRowContainer
	? pageRows.indexOf(parentRowContainer) + 1
	: '_ERROR_';
```
^^^ should be turned into a helper function so it can be reused in the raw html code, which seems to duplicate the landing-page-content code.

Then you can make a helper for finding the containerItem for those things that can be in landing contents or a column with:
```
function getContainerItemInfoForTarget(target) {
   const col = target.closest('[data-eddl-landing-row-col]');
   if (col) {
      const parentColRow = target.closest('[data-eddl-landing-row]');
      const colRows = col.querySelectorAll('[data-eddl-landing-row]');
      const colItemIndex = parentColRow ? colRows.indexOf(parentColRow) + 1 : '_ERROR_';
      return {
       containerItems: colRows.length,
       containerItemsIndex: colItemIndex,
      };
   } else {
     return {
       containerItems: 1,
       containerItemsIndex: 1,
     };
   }
}
```

**Review by bryanpizzillo — APPROVED** [review]

**Review by bennettcc — APPROVED** [review]

**Pull request merged**